Add --download-migration-logs option to generate-script command for gei and ado2gh

synthead opened the pull request:
Blocked by ~~https://github.com/github/gh-gei/pull/371~~.
Blocked by ~~https://github.com/github/octoshift/issues/4702~~.
Blocked by ~~https://github.com/github/octoshift/issues/4751~~.
Closes https://github.com/github/octoshift/issues/4380!

This PR adds `--download-migration-logs` to the `generate-script` command.  When this option is issued, it appends `download-logs` commands after the "wait" commands in the generated scripts.

This PR should ship after https://github.com/github/gh-gei/pull/371 ships!  While this PR doesn't depend on any code from it, shipping this before it would give users an option that isn't supported yet.

- [x] Did you write/update appropriate tests
- [x] Release notes updated (if appropriate)
- [x] Appropriate logging output
- [x] Issue linked
- [x] Docs updated (or issue created) (https://github.com/github/gh-gei/issues/383)

## To-do
- [x] Remove `--wait` and make it the default behavior for `download-logs`
- [x] Add a timeout for checking the log URL (~1 minute should be fine)
- [x] Write to the log and console that the logs are only available for 24 hours
- [x] Add HTTP request debug output for `HttpDownloadService`

<!--
For docs we should review the docs at:
https://docs.github.com/en/early-access/github/migrating-with-github-enterprise-importer
and the README.md in this repo

If a doc update is required based on the changes in this PR, it is sufficient to create an issue and link to it here. The doc update can be made later/separately.

The process to update the docs can be found here:
https://github.com/github/docs-early-access#opening-prs

The markdown files are here: 
https://github.com/github/docs-early-access/tree/main/content/github/migrating-with-github-enterprise-importer
-->

synthead commented:
Should we include this option in the integration tests? :thinking: 

https://github.com/github/gh-gei/blob/018b69c94b018dc0a47a45e226d9e6dff620f0c0/src/OctoshiftCLI.IntegrationTests/GithubToGithub.cs#L50-L50

https://github.com/github/gh-gei/blob/018b69c94b018dc0a47a45e226d9e6dff620f0c0/src/OctoshiftCLI.IntegrationTests/AdoToGithub.cs#L63-L63

synthead commented:
Also, should we add these changes to ado2gh? https://github.com/github/gh-gei/blob/018b69c94b018dc0a47a45e226d9e6dff620f0c0/src/ado2gh/Commands/GenerateScriptCommand.cs

dylan-smith commented:
> Should we include this option in the integration tests? 🤔
> 
> https://github.com/github/gh-gei/blob/018b69c94b018dc0a47a45e226d9e6dff620f0c0/src/OctoshiftCLI.IntegrationTests/GithubToGithub.cs#L50-L50
> 
> https://github.com/github/gh-gei/blob/018b69c94b018dc0a47a45e226d9e6dff620f0c0/src/OctoshiftCLI.IntegrationTests/AdoToGithub.cs#L63-L63

IMO yes!

dylan-smith commented:
> Also, should we add these changes to ado2gh? https://github.com/github/gh-gei/blob/018b69c94b018dc0a47a45e226d9e6dff620f0c0/src/ado2gh/Commands/GenerateScriptCommand.cs

Also yes

synthead commented:
@dylan-smith, sounds good!  I added the option to ado2gh and also added `--download-migration-logs` to the integration tests.

Keep in mind that the integration tests won't pass until https://github.com/github/gh-gei/pull/371 gets merged 😉

dylan-smith commented on src/ado2gh/Commands/GenerateScriptCommand.cs:
```diff
@@ -600,6 +619,11 @@ private void AppendLine(StringBuilder sb, string content)
 
+        private string DownloadMigrationLogScript(string githubOrg, string githubRepo)
+        {
+            return $"./ado2gh download-logs --github-org \"{githubOrg}\" --github-repo \"{githubRepo}\"";
```
I think we should but a conditional in here like all the functions above it, where it checks if _generateScriptOptions.DownloadMigrationLogs is set (then we can remove all the other conditionals used elsewhere)

synthead replied:
Added in b1065614fe55041e29a8ae29edfdc420fc22daea!

I left a conditional for the parallel script generation, since it appends some spaces before the command:

https://github.com/github/gh-gei/blob/b1065614fe55041e29a8ae29edfdc420fc22daea/src/ado2gh/Commands/GenerateScriptCommand.cs#L476-L485

synthead commented:
I slipped a `--wait` option to the download-logs command in this PR since it tackles a race condition related to what's happening in our script.  Happy to move this to another PR if you'd like, but it's small enough that I figured it wouldn't matter too much :slightly_smiling_face: 

I checked our code in Octoshift to see how logs are uploaded, and there is a small gap: the migrations are marked as complete to have a complete log, and then the log file gets uploaded afterwards.  This means that there is a small race condition: it's possible for us to check the status of a migration, see that it is complete, and attempt to download the log before it gets uploaded.

This is where `--wait` comes into play.  Instead of raising an error about the log not existing yet, this option polls the GraphQL endpoint for a new migration URL every ten seconds until it finds a URL to download.  This mitigates this race condition: if we land on it, it'll wait ten seconds and likely download it on the second attempt.  If things are really jammed up in prod, this will let the user download the log whenever it becomes available, too.

In addition, this can be useful to the user, because it can also serve as a "wait for my migration to finish and get the log" without necessarily having to explicitly wait for the migration to finish.  It can also be called in an interactive prompt if users are running migrations by hand without attempting to download multiple times.

However, as of this writing, the integration tests seem to be waiting for that log to populate indefinitely because it appears that uploading the logs to Azure is disabled in prod.  We should probably enable this feature.  It requires a secret to be available to GitHub, but as far as I know, this should be shipped.  Slack thread here: https://github.slack.com/archives/C03BW98SSCA/p1652959517380859

dylan-smith commented on src/OctoshiftCLI.IntegrationTests/TestHelper.cs:
```diff
@@ -585,5 +585,15 @@ public async Task AssertBoardsIntegrationConfigured(string adoOrg, string teamPr
+            _output.WriteLine("Checking that the migration log was downloaded...");
+
+            var migrationLogFile = $"migration-log-{githubOrg}-{repo}.log";
+            var migrationLogExists = await Task.Run(() => File.Exists(migrationLogFile));
+
+            migrationLogExists.Should().Be(true);
```
This can be simplified
```suggestion
            File.Exists(migrationLogFile).Should().Be(true);
```

synthead replied:
I used `Task.Run` here because of this warning: 

> This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.

I added the async `Task.Run` in https://github.com/github/gh-gei/pull/382/commits/ca9e42928c70c02763bed76f6bbc9866e2e73337 in a response to this CI failure:

https://github.com/github/gh-gei/runs/6503637347?check_suite_focus=true

There isn't an async equivalent of `File.Exists` or similar, so I'm wrapping it in a `Task.Run` here :slightly_smiling_face: 

What are your thoughts?

dylan-smith replied:
When this is closer to ready (when prod is updated and int tests are passing) we can pair on this to fix it. But the error is basically saying that you declared this method as async (AssertMigrationLogFileExists), but you're not doing any async stuff in it.  You can fix that in 2 ways, either do some async stuff (which is what you tried to do), or just don't declare the method as async, which I  think is what you should do.

ArinGhazarian replied:
two things:
1. def. don't need to use `Task.Run` 
2. nit: there is a neater syntax for checking booleans: `Should().BeTrue()`

synthead commented:
I also removed the `isHidden` options from `download-logs` introduced in https://github.com/github/gh-gei/pull/396 with https://github.com/github/gh-gei/pull/382/commits/9f5144bee7d57454939159c1477ed222b0883402 :+1:

ArinGhazarian commented on src/ado2gh/Commands/DownloadLogsCommand.cs:
```diff
@@ -60,6 +61,12 @@ HttpDownloadService httpDownloadService
                 Description = "Local file to write migration log to (default: migration-log-ORG-REPO.log)."
             };
 
+            var wait = new Option("--wait")
+            {
+                IsRequired = false,
+                Description = "Waits for log to be ready to download."
+            };
```
Based on the live feedback it'd be better to get rid of `--wait` and always wait with a timeout.

synthead replied:
Added in ed1afe023dff7250306495b522a6a64db736c407!

ArinGhazarian commented on src/ado2gh/Commands/DownloadLogsCommand.cs:
```diff
@@ -127,21 +136,36 @@ HttpDownloadService httpDownloadService
+            string logUrl;
 
-            if (logUrl == null)
+            while (true)
```
Instead of a `while (true)` it's be better to do `while (logUrl is null)` (assuming we shouldn't expect to get an empty or whitespace string). And of course you can add a timeout to the predicate too.

synthead replied:
I replaced the `while(true)` with a for loop in 20b4a78bf1a10bbec8e1cb1e1bbf2ff801b7d7c4.  How's this look?

ArinGhazarian commented on src/ado2gh/Commands/DownloadLogsCommand.cs:
```diff
@@ -127,21 +136,36 @@ HttpDownloadService httpDownloadService
             {
-                throw new OctoshiftCliException($"Migration not found for repository {githubRepo}!");
-            }
+                logUrl = await githubApi.GetMigrationLogUrl(githubOrg, githubRepo);
```
nit: there is a syntactic sugar to throw an exception if the returned value is null:
```suggestion
                logUrl = await githubApi.GetMigrationLogUrl(githubOrg, githubRepo) ?? throw new OctoshiftCliException($"Migration not found for repository {githubRepo}");
```

synthead replied:
Updated in 1a2da58aa27d4397aa73600386cb23f500637225!

ArinGhazarian commented on src/ado2gh/Commands/DownloadLogsCommand.cs:
```diff
@@ -127,21 +136,36 @@ HttpDownloadService httpDownloadService
+
+                if (logUrl == null)
+                {
+                    throw new OctoshiftCliException($"Migration not found for repository {githubRepo}!");
```
nit: IMO the `!` at the end of the error message might be confusing (might be considered as part of the repo name) so I'd just drop it.

synthead replied:
I updated the grammar instead in 7e402a2af69542e9241d83b4076787b2339d5f2f :slightly_smiling_face: How's this look?

https://github.com/github/gh-gei/blob/7e402a2af69542e9241d83b4076787b2339d5f2f/src/gei/Commands/DownloadLogsCommand.cs#L141-L141

ArinGhazarian commented on src/OctoshiftCLI.Tests/gei/Commands/DownloadLogsCommandTests.cs:
```diff
@@ -155,6 +157,62 @@ public async Task Calls_Download_With_Expected_Migration_Log_File()
         }
 
+        [Fact]
+        public async Task Waits_For_URL_To_Populate_When_Wait_Requested()
```
nit: Use `Url` instead of `URL`. More than two letter identifiers should be `Pascal` cases see [this](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions)

synthead replied:
Updated with 5cd1977f6e1d19eff4bffb6c2b3e8a879eb5ab60!

ArinGhazarian commented on src/OctoshiftCLI.Tests/ado2gh/Commands/DownloadLogsCommandTests.cs:
```diff
@@ -155,6 +157,62 @@ public async Task Calls_Download_With_Expected_Migration_Log_File()
         }
 
+        [Fact]
+        public async Task Waits_For_URL_To_Populate_When_Wait_Requested()
```
nit: `Url`

synthead replied:
Updated with 5cd1977f6e1d19eff4bffb6c2b3e8a879eb5ab60!

ArinGhazarian commented on src/OctoshiftCLI.Tests/ado2gh/Commands/DownloadLogsCommandTests.cs:
```diff
@@ -155,6 +157,62 @@ public async Task Calls_Download_With_Expected_Migration_Log_File()
+        public async Task Waits_For_URL_To_Populate_When_Wait_Requested()
+        {
+            // Arrange
+            var githubOrg = "FooOrg";
+            var repo = "foo-repo";
+            var logUrlEmpty = "";
+            var logUrlPopulated = "some-url";
+            var defaultFileName = $"migration-log-{githubOrg}-{repo}.log";
```
would you make these `const`s too. Also regarding `defaultFilename` since it's an interpolated string, not sure if CodeQL supports it yet, C# def. supports making an interpolated string `const` (if possible) now but CodeQL is a bit behind the latest C# lang version so it might complain. I guess you can make the change and see if it does or not 😃

synthead replied:
Updated in 93b135ba09be640054c604d89a7591bf7e40f673!

synthead commented:
Okay @ArinGhazarian and @dylan-smith, I have addressed everyone's concerns, and this is ready for another round of reviews!

Please note that the integration tests will not pass unless the `UPLOAD_MIGRATION_LOGS_ENABLED` feature flag env var is set to `true` in production, and we should definitely ensure that this is enabled before shipping this feature.

synthead commented on src/OctoshiftCLI.IntegrationTests/AdoToGithub.cs:
```diff
@@ -60,7 +60,7 @@ public async Task Basic()
             commitId = await _helper.InitializeAdoRepo(adoOrg, teamProject2, adoRepo2);
             await _helper.CreatePipeline(adoOrg, teamProject2, adoRepo2, pipeline2, commitId);
 
-            await _helper.RunAdoToGithubCliMigration($"generate-script --github-org {githubOrg} --ado-org {adoOrg} --all");
```
Good catch!  Updated with 7c776012c768784c0e587cc1a011b69c9d50b67b!

synthead commented on src/ado2gh/Commands/DownloadLogsCommand.cs:
```diff
@@ -95,8 +98,9 @@ HttpDownloadService httpDownloadService
-            _log.LogInformation($"Downloading logs for organization {githubOrg}...");
+            _log.LogWarning("Migration logs are only available for 24 hours after a migration finishes!");
 
+            _log.LogInformation($"Downloading logs for organization {githubOrg}...");
```
Removed the org name in the log output with 21cb385a659220c5e7caa31611c66e7dffbf15aa!

synthead commented on src/gei/Commands/GenerateScriptCommand.cs:
```diff
@@ -188,6 +195,10 @@ public async Task Invoke(GenerateScriptCommandArgs args)
             }
+            if (args.DownloadMigrationLogs)
+            {
+                _log.LogInformation("DOWNLOAD MIGRATION LOGS: true");
```
Ah you're right!  Added in 9f6e3cd2d6f856faac179554d488bffe48cfce37!

ArinGhazarian commented on src/ado2gh/Commands/DownloadLogsCommand.cs:
```diff
@@ -11,11 +11,15 @@ namespace OctoshiftCLI.AdoToGithub.Commands
         private readonly HttpDownloadService _httpDownloadService;
 
         internal Func<string, bool> FileExists = path => File.Exists(path);
+        internal Func<int, Task> Delay = milliSeconds => Task.Delay(milliSeconds);
```
Why do we need to abstract out the `Delay`? For testing we can set `WaitIntervalInSeconds` and `RetryMaxCount`

ArinGhazarian replied:
Also as a [best practice](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#prefer-asyncawait-over-directly-returning-task) the delegate has to be async and the `Task.Delay` should be awaited:
```c#
internal Func<int, Task> Delay = async milliSeconds => await Task.Delay(milliSeconds);
```

ArinGhazarian commented on src/ado2gh/Commands/DownloadLogsCommand.cs:
```diff
@@ -127,24 +131,29 @@ HttpDownloadService httpDownloadService
-            var logUrl = await githubApi.GetMigrationLogUrl(githubOrg, githubRepo);
 
-            if (logUrl == null)
+            for (var attempt = 1; attempt <= RetryMaxCount; attempt++)
```
I agree, Polly is a better approach here.

ArinGhazarian commented on src/ado2gh/Commands/DownloadLogsCommand.cs:
```diff
@@ -127,24 +131,29 @@ HttpDownloadService httpDownloadService
+                }
 
-            _log.LogSuccess($"Downloaded {githubRepo} log to {migrationLogFile}.");
+                if (attempt == RetryMaxCount)
```
nit: i'd prefer `if (attempt >= RetryMaxCount` just in case

dylan-smith commented:
I just noticed hte Polly changes (and the test refactoring) was only done in the ado2gh command, need to make the same changes to the gei command

dylan-smith reviewed (approved):
Looks good to me! :shipit:
